middle_end/flambda2/types/meet_and_join.ml

```diff
@@ -842,9 +842,46 @@ and meet_product_value_slot_indexed env
     ({ value_slot_components_by_index = components_by_index2 } :
       TG.Product.Value_slot_indexed.t) :
     (TG.Product.Value_slot_indexed.t * TEE.t) Or_bottom.t =
+  let components_by_index1' =
+    Value_slot.Map.fold
+      (fun value_slot (ty, _) components_by_index ->
+        Value_slot.Map.add value_slot ty components_by_index)
+      components_by_index1 Value_slot.Map.empty
+  in
+  let components_by_index2' =
+    Value_slot.Map.fold
+      (fun value_slot (ty, _) components_by_index ->
+        Value_slot.Map.add value_slot ty components_by_index)
+      components_by_index2 Value_slot.Map.empty
+  in
   let<+ components_by_index, env_extension =
-    meet_generic_product env ~components_by_index1 ~components_by_index2
-      ~union:Value_slot.Map.union
+    meet_generic_product env ~components_by_index1:components_by_index1'
+      ~components_by_index2:components_by_index2' ~union:Value_slot.Map.union
+  in
+  let components_by_index =
+    Value_slot.Map.mapi
+      (fun value_slot ty ->
+        let component1 =
+          Value_slot.Map.find_opt value_slot components_by_index1
+        in
+        let component2 =
+          Value_slot.Map.find_opt value_slot components_by_index2
+        in
+        let kind =
+          match component1, component2 with
+          | Some (_, kind1), Some (_, kind2) ->
+            (* CR mshinwell: is this too strong? *)
```

Comment:
In theory, we could return `Bottom` too; but I think we can rely on the invariant that a given value slot can only hold values of a single kind (globally).